netconf_client/ncclient.py

```diff
@@ -21,40 +24,258 @@
 )
 
 
+# Defines the scope for netconf traces
+logger = logging.getLogger("netconf_client_pretty")
+
+
+def pretty_xml(xml):
+    """Reformats a given string containing an XML document (for human readable output)"""
+
+    parser = etree.XMLParser(remove_blank_text=True)
+    tree = etree.fromstring(xml, parser)
+    return etree.tostring(tree, pretty_print=True).decode()
+
+
 class Manager:
-    """A helper class for performing common NETCONF operations
+    """A helper class for performing common NETCONF operations with pretty logging.
 
     This class attempts to be API compatible with the manager object
     from the original ncclient.
 
     This class is also a context manager and can be used with `with`
     statements to automatically close the underlying session.
 
+    NetConf requests and responses are logged using the 'netconf_client' scope.
+    The API functions, which are to be logged, can be configured (default: all but 'close_session').
+    Also the log level can be changed via API (logger.DEBUG is the default).
```

Comment:
This level of configurability seems like overkill. Generally logging config is done with the logger objects, not on the objects doing the logging.

Reply:
I played a little bit with the feature of enabling/disabling some functions.
If all public API functions do some logging, I always saw the "close_session" function to be logged, which was annoying and less helpful. So my first idea was not to include any logging there but then a user might complain that just this function cannot be logged. So I made _all_ functions configurable, whereas all functions are enabled as default but close_session. The logger itself cannot do this because it case no information about functions (ok, we could use logger filters but it seems to be too complicated?).
Other point: When calling the logger, a log level needs to be given (or a certain function, as example logging.debug(), gets called). Consequently, the level would be hard-coded, unless we make it configurable. This was my intention. If you prefer a fixed value, I remove this function and the associated member data, and call logger.debug() instead.

Reply:
Yes I think it's fine for the log level to be fixed to debug. This is how most Python libraries behave.

I'm not sure I understand why `close_session` shouldn't be logged. It's an RPC just like any other, so I don't think it should really be treated specially. I think logging filters are really the correct solution here. You could add an [`extra` dictionary][1] with some more info that would allow the user to filter out certain function calls:

> The third optional keyword argument is extra which can be used to pass a dictionary which is used to populate the __dict__ of the LogRecord created for the logging event with user-defined attributes. These custom attributes can then be used as you like. For example, they could be incorporated into logged messages.

The logger filter receives a `LogRecord` object so it can access these extra parameters and do the filtering there.

[1]: https://docs.python.org/3/library/logging.html#logging.debug

Reply:
One extra remark to `close_session()`: In fact, the `open_session` call is **not** logable because the Manager takes an already connected session. So, we only see a close_session in the log at any time, which seems to be confusing (and spams the klish console when used in the klish-py). 
What is about filters? Hmm, I'm afraid filters won't help because the function name of the API function is not part of the log record (not yet). I need to add it, otherwise a filter could not discard unwanted functions. But adding the function name to the log entry is not very nice. Now I'm a little bit baffled and don't know what is the best...

Reply:
You can add the function name to the `extra` field, which seems to be designed for this exact use case.

Reply:
Sounds awesome, thanks, that's what I'm doing!